File: frontend/src/types/payments.ts
// Copyright (c) 2026 Pablo Health, LLC. Licensed under AGPL-3.0.

/**
 * Self-pay card payment API types.
 *
 * Field names are snake_case to match `app.models.payments` exactly.
 *
 * Nothing here carries a card number, and there is no type that could: the
 * browser posts the card straight to Stripe against a SetupIntent, and the
 * only card-shaped values that come back are the brand, last four digits and
 * expiry the chart renders.
 */

/** The card a practice has on file, as the UI renders it. */
export interface CardOnFileResponse {
  brand: string | null
  last4: string | null
  exp_month: number | null
  exp_year: number | null
  /** False for a setup that was started and never confirmed. */
  chargeable: boolean
}

/**
 * Everything Stripe.js has to be initialised with, from one call.
 *
 * `stripe_account_id` is present only on a deployment whose credentials name
 * an account; it must be passed as `stripeAccount` when it is there and left
 * off entirely when it is not.
 */
export interface CardSetupResponse {
  client_secret: string
  publishable_key: string
  stripe_account_id: string | null
}

/** What a charge sent without an explicit amount would come to. */
export interface ChargeAmountResponse {
  /** `null` when no rate is set — unknown, not free. */
  amount_cents: number | null
  currency: string
}

export interface CreateChargeRequest {
  /** Omit to let the server resolve the amount from `kind`. */
  amount_cents?: number
  appointment_id?: string
  /** Defaults to `session` server-side. */
  kind?: ChargeableKind
}

/**
 * One row of the charge ledger.
 *
 * A decline is a `failed` row returned with HTTP 200, not an error: the
 * attempt happened, and `status_detail` carries the processor's reason.
 */
export interface ChargeResponse {
  id: string
  amount_cents: number
  currency: string
  status: string
  status_detail: string | null
  appointment_id: string | null
  /** What the row IS — see `app.db.models.CHARGE_KINDS`. */
  kind: ChargeKind
  /** The claim a remittance wrote this row out of; `null` otherwise. */
  claim_id: string | null
  write_off_reason: string | null
  note: string | null
  /** Which charge paid this bill off. Provenance, not arithmetic. */
  settled_by_charge_id: string | null
  created_at: string
  updated_at: string | null
}

/**
 * The kinds of thing a ledger row can be.
 *
 * `session` is the practice's own charge for a visit — both the bill and its
 * own payment attempt. `payment` collects against a bill somebody else
 * raised, which a `session` row cannot do without re-billing the amount it
 * settles.
 */
export type ChargeKind =
  | "session"
  | "copay"
  | "payment"
  | "patient_resp"
  | "contractual_adjustment"
  | "write_off"
  | "credit"

/**
 * The two things a card is charged for: the visit at its full rate, and the
 * copay a covered client pays at the door. The rest of the ledger kinds move
 * no money on their own and cannot be raised through the charge route —
 * mirroring the request model's ``Literal["session", "copay"]`` server-side.
 *
 * Derived from `ChargeKind` rather than spelled out again, so a kind that is
 * removed from the ledger cannot survive here as something still chargeable.
 */
export type ChargeableKind = Extract<ChargeKind, "session" | "copay">

/**
 * Why a practice stopped trying to collect. Mirrors
 * `app.db.models.WRITE_OFF_REASONS` — the fixed set the CHECK constraint
 * enforces. `courtesy` and `small_balance` are further gated by practice
 * policy on the billing profile; `hardship` and `error` are not.
 */
export type WriteOffReason = "hardship" | "small_balance" | "courtesy" | "error"

/** A practice-initiated write-off: money it has decided not to collect. */
export interface CreateWriteOffRequest {
  amount_cents: number
  reason: WriteOffReason
  note?: string
}

/** One visit's line of the balance. */
export interface VisitBalanceResponse {
  /** `null` for the rows that hang off no visit, on one trailing line. */
  appointment_id: string | null
  owed_cents: number
  collected_cents: number
  written_off_cents: number
  adjusted_cents: number
  credited_cents: number
  balance_cents: number
}

/**
 * What a client owes, and the arithmetic behind it.
 *
 * `balance_cents` is positive when the client owes the practice and negative
 * when the practice owes the client. A credit is not clamped to zero: a
 * refund the practice owes is exactly what clamping would hide.
 *
 * `outcome_known` is false when this client's payer sends its remittances
 * somewhere other than Pablo. The arithmetic is still right over the rows we
 * have; what it cannot include is the client's share of an insured visit,
 * which only an 835 tells us. Rendering the total without saying so states a
 * settled account that was never settled.
 */
export interface BalanceResponse {
  owed_cents: number
  collected_cents: number
  written_off_cents: number
  adjusted_cents: number
  credited_cents: number
  balance_cents: number
  outcome_known: boolean
  by_visit: VisitBalanceResponse[]
}

/** One client on the practice-wide balances list. */
export interface ClientBalanceItem {
  patient_id: string
  patient_name: string
  balance_cents: number
  currency: string
  /** When the money behind the balance first went on the ledger. */
  outstanding_since: string
  /**
   * False when this client's payer sends its remittances elsewhere, so what
   * is owed is at least this and possibly more.
   */
  outcome_known: boolean
}

export interface BalancesResponse {
  items: ClientBalanceItem[]
}

File: backend/app/models/payments.py
# Copyright (c) 2026 Pablo Health, LLC. Licensed under AGPL-3.0.

"""Domain and API models for self-pay card payments.

Two domain models — the card a practice has on file for a client, and one row
of the charge ledger — plus the request/response shapes the API speaks.

Neither domain model carries a card number, because nothing in the system ever
holds one: the browser posts the card straight to Stripe and hands the backend
an opaque payment-method id. ``brand``/``last4``/``exp_*`` are display fields,
and they are the only card-shaped values that exist anywhere below this line.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

#: Largest single charge the API will attempt, in minor units. Not a processor
#: limit — a blast-radius cap on a fat-fingered amount (cents typed into a
#: dollars box, or a UI regression sending 100x). Far above any plausible
#: single self-pay session and far below "career-ending mistake". Rejected with
#: a 422, never silently clamped.
MAX_CHARGE_CENTS = 1_000_000


class CardOnFile(BaseModel):
    """The card a practice keeps on file for one client.

    ``stripe_payment_method_id`` is ``None`` between "setup started" and "the
    browser confirmed the card" — that row is not chargeable, which is what
    :attr:`chargeable` reports.
    """

    id: str
    patient_id: str
    stripe_customer_id: str
    stripe_payment_method_id: str | None = None
    card_brand: str | None = None
    card_last4: str | None = None
    card_exp_month: int | None = None
    card_exp_year: int | None = None

    @property
    def chargeable(self) -> bool:
        return bool(self.stripe_payment_method_id)


class PatientCharge(BaseModel):
    """One row of the charge ledger — one attempt to charge a client's card."""

    id: str
    patient_id: str
    appointment_id: str | None = None
    # What the row is — see ``app.db.models.CHARGE_KINDS``. Defaulted so a
    # ledger row read from before the column existed, or written by a caller
    # that predates it, is the full-rate session charge it always was.
    kind: str = "session"
    claim_id: str | None = None
    write_off_reason: str | None = None
    note: str | None = None
    settled_by_charge_id: str | None = None
    amount_cents: int
    currency: str
    status: str
    status_detail: str | None = None
    stripe_payment_intent_id: str | None = None
    created_by_user_id: str
    created_at: datetime
    updated_at: datetime | None = None


class CardSetupResponse(BaseModel):
    """What the browser needs in order to collect a card.

    All three fields configure the same Stripe.js instance, which is why they
    are returned together rather than assembled from separate sources: the
    publishable key and the account have to match the secret key the resulting
    card will eventually be charged with, and a mismatch produces a card that
    saves and then never charges.

    ``publishable_key`` is public by design — it is the key the browser posts
    card details with, and it does nothing else.

    ``stripe_account_id`` is present only when the deployment's credential
    provider named an account; Stripe.js has to be initialised with it as
    ``stripeAccount`` in that case, and must not be in the default one.
    """

    client_secret: str
    publishable_key: str
    stripe_account_id: str | None = None


class CardSetupConfirmation(BaseModel):
    """The SetupIntent the browser just confirmed.

    Only the id. What card actually got attached is then read back from
    Stripe — the browser is not trusted for the display fields, because a
    caller that could write them could make the stored card read as one card
    while a charge went to another.
    """

    setup_intent_id: str = Field(min_length=1, max_length=255)


class CardOnFileResponse(BaseModel):
    """The card on file, as the UI renders it. Display fields only."""

    brand: str | None = None
    last4: str | None = None
    exp_month: int | None = None
    exp_year: int | None = None
    chargeable: bool = False


class CreateChargeRequest(BaseModel):
    """A one-click charge.

    ``amount_cents`` is optional: left out, the amount is resolved on the
    server from what ``kind`` says this charge is. Sending it overrides that
    for this one charge (a partial payment, a late-cancellation fee, or a
    copay nobody has on file).

    ``kind`` is the two things a card is charged for. ``session`` resolves to
    the client's own rate, falling back to the appointment type's default
    fee; ``copay`` resolves to what a covered client pays at the door and
    records a row a later remittance can net out. The remaining ledger kinds
    are not charges at all — a write-off or a contractual adjustment moves no
    money and cannot be raised here.

    Currency is not a parameter — the deployment charges in one currency and a
    caller cannot pick another.
    """

    amount_cents: int | None = Field(default=None, gt=0, le=MAX_CHARGE_CENTS)
    appointment_id: str | None = None
    kind: Literal["session", "copay"] = "session"


class CreateWriteOffRequest(BaseModel):
    """A practice-initiated write-off: money it has decided not to collect.

    ``reason`` is checked at the route against ``app.db.models.WRITE_OFF_REASONS``
    — the same set the CHECK constraint enforces — rather than encoded as a
    ``Literal`` here, so there is one list to update rather than two that can
    drift apart. ``courtesy`` and ``small_balance`` are further gated by
    practice policy; ``hardship`` and ``error`` are not.
    """

    amount_cents: int = Field(gt=0, le=MAX_CHARGE_CENTS)
    reason: str = Field(min_length=1, max_length=24)
    note: str | None = Field(default=None, max_length=2000)


class ChargeAmountResponse(BaseModel):
    """What a charge sent without an explicit amount would come to.

    Exists so the clinician sees the figure *before* committing to it. The
    charge route resolves the same value from the same helper; asking for it
    first is a read, and reading it changes nothing.

    ``amount_cents`` is ``None`` when neither the client nor the appointment
    type sets a rate. That is not zero and must not be rendered as free — it
    means the UI has to ask for an amount, exactly as the charge route would
    refuse without one.
    """

    amount_cents: int | None = None
    currency: str


class ChargeResponse(BaseModel):
    """One ledger row, as the practice sees it.

    No Stripe customer or payment-method id and no card data: the ledger is
    amounts, statuses and reasons.
    """

    id: str
    amount_cents: int
    currency: str
    status: str
    status_detail: str | None = None
    appointment_id: str | None = None
    kind: str = "session"
    claim_id: str | None = None
    write_off_reason: str | None = None
    note: str | None = None
    settled_by_charge_id: str | None = None
    created_at: datetime
    updated_at: datetime | None = None


class VisitBalanceResponse(BaseModel):
    """One visit's line of the balance. ``appointment_id`` is ``None`` for the
    rows that hang off no visit, collapsed into a single trailing line."""

    appointment_id: str | None = None
    owed_cents: int
    collected_cents: int
    written_off_cents: int
    adjusted_cents: int
    credited_cents: int
    balance_cents: int


class BalanceResponse(BaseModel):
    """What a client owes, and the arithmetic that produced it.

    ``balance_cents`` is positive when the client owes the practice and
    negative when the practice owes the client — a credit is not clamped to
    zero, because a refund the practice owes is exactly the thing a clamped
    balance would hide.
    """

    owed_cents: int
    collected_cents: int
    written_off_cents: int
    adjusted_cents: int
    credited_cents: int
    balance_cents: int
    by_visit: list[VisitBalanceResponse]


class ClientBalanceItem(BaseModel):
    """One client on the practice-wide balances list.

    ``outstanding_since`` is when the money behind the balance first went on
    the ledger, which is what the list is ordered by: the oldest balance is
    the one that most needs a conversation, not the largest.
    """

    patient_id: str
    patient_name: str
    balance_cents: int
    currency: str
    outstanding_since: datetime


class BalancesResponse(BaseModel):
    """Every client carrying a balance, oldest first."""

    items: list[ClientBalanceItem]
